controller_interface/src/chainable_controller_interface.cpp

```diff
@@ -44,53 +44,136 @@ return_type ChainableControllerInterface::update(
-        "correct and recompile the controller with name '%s' and try again.",
-        interface.get_name().c_str(), get_node()->get_name());
-      state_interfaces.clear();
-      break;
+      std::string error_msg =
+        "The prefix of the interface '" + interface.get_prefix_name() +
+        "' does not equal the controller's name '" + get_node()->get_name() +
+        "'. This is mandatory for state interfaces. No state interface will be exported. Please "
+        "correct and recompile the controller with name '" +
+        get_node()->get_name() + "' and try again.";
+      throw std::runtime_error(error_msg);
+    }
+    auto state_interface = std::make_shared<hardware_interface::StateInterface>(interface);
+    const auto interface_name = state_interface->get_name();
+    auto [it, succ] = exported_state_interfaces_.insert({inteface_name, state_interface});
+    // either we have name duplicate which we want to avoid under all circumstances since interfaces
+    // need to be uniquely identify able or something else really went wrong. In any case abort and
+    // inform cm by throwing exception
+    if (!succ)
+    {
+      std::string error_msg =
+        "Could not insert StateInterface<" + inteface_name +
+        "> into exported_state_interfaces_ map. Check if you export duplicates. The "
+        "map returned iterator with interface_name<" +
+        it->second->get_name() +
+        ">. If its a duplicate adjust exportation of InterfacesDescription so that all the "
+        "interface names are unique.";
+      exported_state_interfaces_.clear();
+      exported_state_interface_names_.clear();
+      state_interfaces_ptrs_vec.clear();
+      throw std::runtime_error(error_msg);
     }
+    ordered_exported_state_interfaces_.push_back(state_interface);
+    exported_state_interface_names_.push_back(interface_name);
+    state_interfaces_ptrs_vec.push_back(state_interface);
   }
 
-  return state_interfaces;
+  return state_interfaces_ptrs_vec;
 }
 
-std::vector<hardware_interface::CommandInterface>
+std::vector<std::shared_ptr<hardware_interface::CommandInterface>>
 ChainableControllerInterface::export_reference_interfaces()
 {
   auto reference_interfaces = on_export_reference_interfaces();
+  std::vector<std::shared_ptr<hardware_interface::CommandInterface>> reference_interfaces_ptrs_vec;
+  reference_interfaces_ptrs_vec.reserve(reference_interfaces.size());
+  exported_reference_interface_names_.reserve(reference_interfaces.size());
+  ordered_reference_interfaces_.reserve(reference_interfaces.size());
+
+  // BEGIN (Handle export change): for backward compatibility
+  // check if the "reference_interfaces_" variable is resized to number of interfaces
+  if (reference_interfaces_.size() != reference_interfaces.size())
+  {
+    // TODO(destogl): Should here be "FATAL"? It is fatal in terms of controller but not for the
+    // framework
```

Comment:
If we decided to throw then this comment might go away.

Reply:
```suggestion
``` 
correct! removing